### DISCUSSION TITLE
Need help getting set up

### DISCUSSION THREAD

**Discussion by ssicari** [Robotic-Decision-Making-Lab/blue#45] [Q&A] [answered]
Hello,
When I run ros2 launch blue_bringup bluerov2_heavy.launch.py I get 
```bash
ER: broadcast request timeout, retries left 4
[mavros_node-1] [WARN] [1688668288.372672914] [mavros.sys]: VER: broadcast request timeout, retries left 3
[mavros_node-1] [WARN] [1688668289.373849672] [mavros.sys]: VER: unicast request timeout, retries left 2
[mavros_node-1] [WARN] [1688668290.372700514] [mavros.sys]: VER: unicast request timeout, retries left 1
[mavros_node-1] [WARN] [1688668291.373079567] [mavros.sys]: VER: unicast request timeout, retries left 0
[mavros_node-1] [WARN] [1688668292.373835544] [mavros.sys]: VER: your FCU don't support AUTOPILOT_VERSION, switched to default capabilities
```
I am also getting a lot of plugins ignored so maybe that is related to this but, I am unsure why this is and I am looking into it, just curious if you might know what is going on. Thank you!

**Comment by evan-palmer** [comment]
Thanks for checking the project out @ssicari! I apologize for the lack of documentation. I haven't created an official release yet, so things are a bit unstable at the moment. With regard to the installation documentation, that is not in a completed state (see #34) and I anticipate that the issues that you are experiencing are related to that. What development environment are you using? Also what GPU drivers are you using? You may have more success with the docker environment which is set up to support usage as a [dev container](https://code.visualstudio.com/docs/devcontainers/create-dev-container). That is the main environment that I have used for testing and development.

**Reply by ssicari** [reply]
Hello, 
Don't worry about the documentation this is a big project and it's still early so I understand. As for the see #34 I do have gazebo ardusub and ardupilot gazebo already installed, I have been using Clyde's repo for some stuff and they seem to work there. My development environment is Ubuntu 22.04 with ROS2 Humble, ArduSub, ardupilot_gazebo, and gazebo garden. My GPU drivers are 
Mesa 22.2.5-0ubuntu0.1~22.04.3
The main problem I seem to be getting now is a connection refusal to mavros_router, so I will look into that and  I will look into the docker environment and see if I have any luck setting that up. Thank you!

Here is an output of part of the mavros refusal
![image](https://github.com/evan-palmer/blue/assets/74316868/0c4a5d74-e5a6-4ace-bcb1-8a83fb3fe1ab)

**Reply by ssicari** [reply]
Hello,
I am trying to setup the dev environment and run the Dockerfile, but I keep getting this error.
![image](https://github.com/evan-palmer/blue/assets/74316868/48cfde63-11d8-4fdf-a643-5df0a110f014)
I am working on looking into it but was wondering if you have seen anything like this?
Thank you

**Reply by evan-palmer** [reply]
I apologize for the late reply here--I have been away for a few days! I haven't yet experienced this issue. The good news is that the image fully builds. My guess is that there is an issue with the graphics driver. When integrating support for visualization from Docker, I made the decision to start with support for the Xorg open source GPU drivers to ensure that folks with either NVIDIA GPUs or without could use the project. Could you try switching to the xorg driver to see if that works?

**Reply by evan-palmer** [reply]
With regards to installing natively, I would be curious to see what the launch log looks like. When I experience that issue, it is typically when a plugin fails to start. My guess would be that #25 is rearing its ugly head again (I have been putting off fixing that bug while I work to meet a deadline)

**Reply by ssicari** [reply]
If these are the logs you are talking about for the native install after I try to run 

```bash
ros2 launch blue_bringup bluerov2_heavy.launch.py
```

these are them. As for the xorg drivers, I was trying to avoid installing them on my host machine so I tried to add the line 

```bash
RUN apt-get update && apt-get install -y xserver-xorg-video-all
```

at the top of the Dockerfile. I imported the devcontainer file and Dockerfile to a new dir on my host machine that I have a .devcontainer folder in. When trying to start the dev container in VS Code now though I encounter an error to pick a keyboard layout but can not interact with the terminal. I am not very familiar with dev containers so I will need to look into this more tomorrow. I also plan to look at the revised changes you mentioned tomorrow as well. Anything else you need please let me know, thank you!


[events.log](https://github.com/evan-palmer/blue/files/12021348/events.log)
[logger_all.log](https://github.com/evan-palmer/blue/files/12021349/logger_all.log)

**Reply by evan-palmer** [reply]
If you would like to avoid installing the Xorg drivers (totally understandable), something that you could also try is adding a new stage to the Dockerfile which configures the image to support the NVIDIA drivers. [Here](https://github.com/clydemcqueen/orca4/blob/cff6caa72c084dca7bfcf0f179e65cc2f985c958/docker/Dockerfile#L35) is the layer from Clyde's repo. I believe that he got this from [this](https://github.com/athackst/vscode_ros2_workspace/tree/humble-nvidia/.devcontainer) repository which should have all of the steps needed to setup for the NVIDIA drivers. The Dockerfile used in that repo is [here](https://github.com/athackst/dockerfiles/blob/main/ros2/humble-cuda.Dockerfile). One potential way for you to test this quickly could be to add a new stage right before the `develop` stage like this:

```bash
FROM sim AS nvidia

# The following is from the athackst Dockerfile that I linked
RUN apt-get update \
 && apt-get install -y -qq --no-install-recommends \
  libglvnd0 \
  libgl1 \
  libglx0 \
  libegl1 \
  libxext6 \
  libx11-6

ENV NVIDIA_VISIBLE_DEVICES all
ENV NVIDIA_DRIVER_CAPABILITIES graphics,utility,compute
ENV QT_X11_NO_MITSHM 1

FROM nvidia AS develop

...
```
From here, you would then update the dev container to use the proper arguments as is done by the athackst `devcontainer.json`. I apologize that this has been painful. 

With regards to the log file, would you be willing to share the `launch.log` file? The filepath should be documented right at the beginning of the launch trace (something like `/home/ssicari/.ros/log/...`)

**Reply by ssicari** [reply]
Thank you I will try to update the Dockerfile before I natively install the xorg drivers. In regards to the launch.log file I believe this should be it.
[launch.log](https://github.com/evan-palmer/blue/files/12029972/launch.log)

**Reply by ssicari** [reply]
I just sent the changes please let me know if anything needs more revisions. I believe from Clyde's comment he wanted to keep the same file names, and just change the world names. If I misunderstood that just let me know and I can revert the world names back to having the underwater suffix. I am going to start working on the Dockerfiles to see if I can get that running. Thank you!

**Reply by ssicari** [reply]
I realized that the graphic drivers I have are mesa and not Nvidia so I do not believe modifying the dockerfile to handle Nvidia drivers will work. I am working on installing the xorg drivers natively and believe I have them installed but when I run the command "glxinfo | grep "OpenGL renderer" it still says the mesa drivers are in use. I am struggling to find a solution right now. Do you happen to know anything about the xorg drivers? I am going to attach the failed docker log file from before as well, thank you for all the help! 
[remoteContainers-2023-07-13T20-12-01.114Z.log](https://github.com/evan-palmer/blue/files/12044222/remoteContainers-2023-07-13T20-12-01.114Z.log)

**Reply by evan-palmer** [reply]
When you open the "Additional Drivers" application in Ubuntu, do you have the "X.Org X server" driver selected? This should look something similar to the following:

![image](https://github.com/evan-palmer/blue/assets/75103118/c583576f-9948-4d8b-a17f-be7bc877e120)

Also, could you share the output of the following command (this will display the driver loaded):

```bash
sudo lshw -c video | grep driver
```

In my case, the output of the above is as follows:

```bash
configuration: depth=32 driver=nouveau latency=0 resolution=1920,1080
```

**Reply by ssicari** [reply]
Hello,
This is my output to the command:
```configuration: depth=32 driver=i915 latency=0 mode=1920x1200 resolution=1920,1200 visual=truecolor xres=1920 yres=1200```
As for the additional driver section, mine looks like this
![image](https://github.com/evan-palmer/blue/assets/74316868/8e784e65-b0e0-4352-a5d2-38b28b076e5f)
When I run the command:
```dpkg -l xserver-xorg-video-intel```
I get this, which I believe the "ii" means it is installed
![image](https://github.com/evan-palmer/blue/assets/74316868/34823e95-075f-4ef6-8fb8-575244fc7ba5)

**Reply by evan-palmer** [reply]
Oh interesting. That's the same driver that I am using on my laptop (I do most of my development on my desktop which is where the Xorg driver is installed). I'll do some debugging on there. In the meantime, would you be willing to share the VSCode log when dev container launch fails? After it fails, it will give you the option to open the devcontainer.json file locally. Select that option, then the log file of interest should open automatically

Update: I tested out the project on my laptop which has the same drivers that your system has and was able to start the development container and project. I will be curious to see what the dev container log says

**Comment by ssicari** [comment]
The additional drivers section is completely empty for me. I installed xserver-xorg-intel earlier and it says on the terminal it was installed. I can get you the output of that command either later tonight or early tomorrow morning. Thank you!On Jul 13, 2023, at 15:49, Evan Palmer ***@***.***> wrote:﻿
When you open the "Additional Drivers" application in Ubuntu, do you have the "X.Org X server" driver selected? This should look something similar to the following:

Also, could you share the output of the following command (this will display the driver loaded):
sudo lshw -c video | grep driver
In my case, the output of the above is as follows:
configuration: depth=32 driver=nouveau latency=0 resolution=1920,1080

**Comment by ssicari** [comment]
Hello,I’m going to be unavailable til Wednesday so unfortunately I can’t share the log file until I get back. I will share it as soon as I get back and go from there. I apologize for the inconvenience and appreciate the help!Thank youOn Jul 16, 2023, at 20:35, Evan Palmer ***@***.***> wrote:﻿
Oh interesting. That's the same driver that I am using on my laptop. I'll do some debugging on there. In the meantime, would you be willing to share the VSCode log when dev container launch fails? After it fails, it will give you the option to open the devcontainer.json file locally. Select that option, then the log file of interest should open automatically

**Comment by evan-palmer** [comment]
That's no problem at all! I appreciate your patience. I am sorry that this has been painful!

**Comment by evan-palmer** [comment]
I also forgot to ask this earlier, what command did you use to launch the system? That original error that you got with your native install can happen if ArduSub isn't available. ArduSub is only launched when the simulation is also launched:

```
ros2 launch blue_bringup bluerov2_heavy.launch.py use_sim:=true
```

**Reply by ssicari** [reply]
Hello, 
Sorry for the delayed response! Here is my log file: 
[remoteContainers-2023-07-20T16-14-50.090Z.log](https://github.com/evan-palmer/blue/files/12111852/remoteContainers-2023-07-20T16-14-50.090Z.log)
As for the command I was running, I was originally running the base command where I got the previous error. I have run the sim command that you listed above and gazebo does open. Is there a way to control it that way however? I can't seem to get it to run with Qgroundcontrol or through the terminal I start it in. Is there anyway to test the sim to see if it is working?
Thank you, I appreciate all the help!

**Reply by evan-palmer** [reply]
I'm glad to hear that it launched! As for QGC, that's strange. I haven't encountered that problem before. I am able to launch the simulation and QGC establishes a connection with the SITL interface. When you look at the connections tab in QGC do you see any active connections? 

If you would like to test it without QGC, you can use some of the scripts in the `blue_utils/scripts/` directory. They are intended more for debugging and testing out custom controllers, but here is what each does and how you can modify them for testing:

- `arm.sh`: Enables PWM passthrough mode and arms the custom controller (in this case, just the ISMC). Passthrough mode should be enabled while in ArduSub manual mode.
- `disarm.sh`: Disables PWM passthrough mode and disarms the custom controller
- `send_cmd.sh`: Can be used to send a velocity command to the controller (modify this according to your needs)
- `send_pwm.sh`: Can be used to send a PWM command; if passthrough mode is enabled, this will control a single thruster, if passthrough mode is disabled, but ArduPilot is armed, then this will enable sending commands similar to the MAVProxy CLI

Another method is to use the [teleop_twist_keyboard](https://index.ros.org/r/teleop_twist_keyboard/#rolling) package (included as a rosdep) and launch that to integrate with the custom controller:

```bash
>$ ros2 launch blue_bringup bluerov2_heavy.launch.py use_sim:=true

# In a separate tab
>$ ./blue_utils/scripts/arm.sh
>$ ros2 run teleop_twist_keyboard teleop_twist_keyboard --ros-args -r /cmd_vel:=/blue/ismc/cmd_vel
```

**Reply by evan-palmer** [reply]
With regards to the dev container log, that's strange; I haven't had any issues with ArduPilot not installing. The user is changed in [this](https://github.com/evan-palmer/blue/blob/b4c86cce8901a76908399a15c3eeeb7fa71cf0a1/.docker/Dockerfile#L112) line. If you are still using the dev container, could you try installing with the latest version of main? If you have switched over to just using your native installation though, don't worry about it

**Reply by ssicari** [reply]
It does seem to establish a connection with in qgroundcontrol, but there is no rov icon and when I try to do any manual path planning it does not seem to work or arm the rov in the sim. I will also look at the other ways you mentioned to test it out. I also plan on looking on testing out ROS with it, are there any ROS files already created with basic movement to test out the sim or should I make my own? As for the dev container I am going to try to test out the native install for now, if I can get that working I will be able to ignore the dev container. Thank you so much for the help!

**Reply by evan-palmer** [reply]
Oh interesting. I'll admit that I haven't tested QGC extensively with the project so there could certainly be a bug in the configuration that I haven't discovered. With regards to the ROS files made available to test things out, the `blue_control` package provides an interface for developing custom controllers, and an ISMC has been implemented which you can send velocity commands to. The `teleop_twist_keyboard` sends commands to that controller. You can also develop your own custom velocity controllers there. I have not implemented any motion or path planning, however. I am also planning to implement a more robust interface for GUIDED mode in the next week or so, but it doesn't exist at the moment: I'm waiting on a PR (checking on link) before I fully implement that so that I can continue to use the velocity command interface.

**Reply by evan-palmer** [reply]
Another thing that you could do to test out ROS integration is sending setpoint position commands to MAVROS with the ROV set to GUIDED mode. I don't have any examples of that in the project but a simple publisher should be able to do that for you

**Reply by ssicari** [reply]
Ah I see, the teleop commands are working. For now I have 2 more quick questions. I am struggling to find the configuration file that handles the Qgroundcontrol part, I'd be more than happy to check that out and play around with it. As well as how do you change the ardusub mode from manual to guided? I am also unsure if the send scripts are not working or if I am just using them incorrectly, would you be able to show me how to properly use them as you did with the teleop command? I apologize for the all the questions, Thank you!

**Reply by evan-palmer** [reply]
So the main places that I would look for configuring QGC are the `blue_description/config/mavros.yaml` file (warning the documentation for MAVROS is cheeks) and the Gazebo world file. The PX4 SITL interface may be able to point you in the right direction as well. They have done a good job of documenting their stuff and building out their simulation. My initial hypothesis as to why the location isn't showing up is that there may be a conflict with the Gazebo localization interface which publishes in the local frame and what QGC is expecting (probably world).

In terms of the CLI, there isn't a great interface for interacting with MAVROS other than through the ROS CLI. That is what inspired me to make some of those bash scripts (e.g., `send_pwm.sh` interacts with MAVROS using the ROS 2 CLI). A number of the interfaces for arming and changing modes are made available as services through MAVROS. In the distant future, I am hoping to get this project integrated as a plugin with BlueOS to provide a front-end for some of those things.

**Reply by ssicari** [reply]
Thank you I appreciate the response. I will play around with that more tonight and tomorrow!

**Reply by ssicari** [reply]
> send_pwm.sh can be used to send a PWM command; if passthrough mode is enabled, this will control a single thruster, if 
> passthrough mode is disabled, but ArduPilot is armed, then this will enable sending commands similar to the MAVProxy CLI"

Hello, when you say this is passthrough mode also controled by a ROS CLI command? Also do you have to change the script to run it everytime for different thrusters? When you say it will enable sending commands similar to MAVProxy CLI does that mean by changing the script or by sending the command from the script on the CLI? Right now I have to stop running the script and then change the values and rerun the script for new movement, I was unsure if that was the only way other than sending the commands directly on the CLI, thank you

Also this is weird but I can find the mavros.yaml file on the github but not in my local repo
![image](https://github.com/evan-palmer/blue/assets/74316868/5fe4d039-d483-4d6f-93e4-1d7c19803a6e)
This is the screenshot. I am trying to figure out why that is or if I am missing something simple

**Reply by evan-palmer** [reply]
> Hello, when you say this is passthrough mode also controled by a ROS CLI command?

Passthrough mode is managed by the `blue_manager` node which you can send commands to using the service interface. This mode isn't available through ArduSub. 

> Also do you have to change the script to run it everytime for different thrusters? When you say it will enable sending commands similar to MAVProxy CLI does that mean by changing the script or by sending the command from the script on the CLI? Right now I have to stop running the script and then change the values and rerun the script for new movement, I was unsure if that was the only way other than sending the commands directly on the CLI, thank you

Yeah, you will have to change the script to modify it each time. Those scripts are really just intended for debugging (similar to the ROS CLI). The system is designed to take inputs from a publisher more than from a command line interface. A few alternatives to the keyboard inputs, CLI, or publisher could be to make a simple QT GUI or to integrate one of the existing joystick ROS nodes to support sending commands from an XBox controller or something similar.

> Also this is weird but I can find the mavros.yaml file on the github but not in my local repo

It looks like you have a version from a few commits ago (see #49). I moved some things around in the latest version to help support folks that want to create their own custom models. In the version that you have, the configuration should be in the `blue_bringup/config` directory.

**Reply by ssicari** [reply]
Ah I see I thought I checked earlier and had the most updated branch but I did not, I pulled the most recent version, thank you for the information on the controls as well!

**Reply by ssicari** [reply]
Hello, Sorry to bother you again. After further investigation qgroundcontrol recognizes the motors being armed and disarmed. It still misses the icon, and when I try to run a basic way point mission it failed because it can not enter guided mode. I have tried to manually test if I can put the vehicle into guided mode via the ROS CLI but can not get it to recognize the command. Following the mavros docs I can check the current mode it is in and see the possible modes it could change to but everytime I try to actually change the mode it says invalid passed service type. I was wondering if you have manually changed modes before and had a working command because the ones directly from the docs don't seem to work along with a few other variations I have tried. Thank you!

**Reply by evan-palmer** [reply]
I've split this discussion into a new discussion [here](#57), and I'm going to close this now to make it easier for folks go through in the future.

**Comment by evan-palmer** [comment, marked as answer]
I'm writing this as a separate comment for future readers, the error seen in this initial comment is caused when MAVROS is unable to establish a connection with ArduSub. This can happen for a number of reasons, but in this particular case, the command

```bash
ros2 launch blue_bringup bluerov2_heavy.launch.py
```

does not launch ArduSub. At the time of writing, ArduSub is launched only when the Gazebo SITL interface is started:

```bash
ros2 launch blue_bringup bluerov2_heavy.launch.py use_sim:=true
```

I have opened #51 to improve the documentation regarding this.

**Reply by SusannaMunari** [reply]
Hello,

I am encountering the same issue when running this:
`ros2 launch blue_bringup bluerov2_heavy.launch.yaml use_sim:=true`

I get this output, and I am not able to connect and simply echo the topics. 
![image](https://github.com/Robotic-Decision-Making-Lab/blue/assets/25926194/41f49262-6154-42fd-b601-519ed4af6a43)


Curiously, the same process works smoothly when I try with the normal configuration:
`ros2 launch blue_bringup bluerov2.launch.yaml use_sim:=true`

With this I am connected, I can access the topics and change flight mode. 

Is there any configuration that I am missing?

**Reply by evan-palmer** [reply]
Hi @SusannaMunari, I've made a new discussion [here](https://github.com/Robotic-Decision-Making-Lab/blue/discussions/191) for your question
